src/Chapter3.hs

```diff
@@ -480,6 +528,89 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle = Castle
+  {
+    name :: String
+  }
+
+data IsCastle = 
+  Yes Castle |
+  No
+
+data WhichBuilding = 
+  Chruch |
+  Library
+
+data People = 
+  One Int |
+  Two Int |
+  Three Int |
+  Four Int
+
+data House = House
+  {
+    people :: People
+  }
+
+-- How to represent the Wall object?
+data City = City
+  {
+    castle :: IsCastle,
+    wall :: Bool,
```

Comment:
This is the correct data type, that can represent what we have in requirements 👍🏼 
However, this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This is again "Make illegal states unrepresentable" pattern 🙂 

There are few options on how to resolve this, you can use any you like!